# Exercice 4 — Introduction à pandas

**Durée cible : 45 minutes**  
**Niveau : Junior**

---

## 💡 Tips

- `pd.DataFrame(liste_de_dicts)` — crée un DataFrame depuis une liste de dicts
- `.head(n)` affiche les n premières lignes (5 par défaut)
- `.info()` → types des colonnes + valeurs nulles / `.describe()` → stats descriptives
- `.shape` → `(nb_lignes, nb_colonnes)` — c'est un attribut, pas une méthode (pas de `()`)
- Sélection d'une colonne : `df["colonne"]` → Series / `df[["col1", "col2"]]` → DataFrame
- Filtre : `df[df["colonne"] > valeur]` — la condition va **entre crochets**
- Plusieurs conditions : `(cond1) & (cond2)` avec des parenthèses autour de chaque condition
- Nouvelle colonne : `df["nouvelle_col"] = df["col1"] * df["col2"]`
- `.groupby("col").agg({"col2": "sum", "col3": "mean"})` — agrège plusieurs colonnes en une fois
- `.sort_values("col", ascending=False)` — trie par ordre décroissant

---


## Dataset

Exécute cette cellule pour créer le DataFrame de travail.


In [1]:
import pandas as pd

data = [
    {"vendeur": "Alice", "region": "Nord", "produit": "Laptop",  "categorie": "Tech",     "quantite": 3,  "prix_unitaire": 1200, "mois": "Janvier"},
    {"vendeur": "Bob",   "region": "Sud",  "produit": "Chaise",  "categorie": "Mobilier", "quantite": 10, "prix_unitaire": 150,  "mois": "Janvier"},
    {"vendeur": "Alice", "region": "Nord", "produit": "Ecran",   "categorie": "Tech",     "quantite": 5,  "prix_unitaire": 400,  "mois": "Février"},
    {"vendeur": "Clara", "region": "Est",  "produit": "Bureau",  "categorie": "Mobilier", "quantite": 2,  "prix_unitaire": 600,  "mois": "Janvier"},
    {"vendeur": "Bob",   "region": "Sud",  "produit": "Clavier", "categorie": "Tech",     "quantite": 8,  "prix_unitaire": 80,   "mois": "Février"},
    {"vendeur": "Clara", "region": "Est",  "produit": "Laptop",  "categorie": "Tech",     "quantite": 2,  "prix_unitaire": 1200, "mois": "Février"},
    {"vendeur": "Alice", "region": "Nord", "produit": "Lampe",   "categorie": "Mobilier", "quantite": 15, "prix_unitaire": 45,   "mois": "Mars"},
    {"vendeur": "Bob",   "region": "Sud",  "produit": "Ecran",   "categorie": "Tech",     "quantite": 3,  "prix_unitaire": 400,  "mois": "Mars"},
    {"vendeur": "Clara", "region": "Est",  "produit": "Chaise",  "categorie": "Mobilier", "quantite": 6,  "prix_unitaire": 150,  "mois": "Mars"},
    {"vendeur": "Alice", "region": "Nord", "produit": "Clavier", "categorie": "Tech",     "quantite": 4,  "prix_unitaire": 80,   "mois": "Mars"},
    {"vendeur": "Bob",   "region": "Sud",  "produit": "Bureau",  "categorie": "Mobilier", "quantite": 1,  "prix_unitaire": 600,  "mois": "Janvier"},
    {"vendeur": "Clara", "region": "Est",  "produit": "Lampe",   "categorie": "Mobilier", "quantite": 20, "prix_unitaire": 45,   "mois": "Février"},
]

df = pd.DataFrame(data)


---

## Partie 1 — Inspection du DataFrame

### 1.1 — Premiers regards

1. Affiche les **5 premières lignes** du DataFrame
2. Affiche le **nombre de lignes et de colonnes** (utilise `.shape`)
3. Affiche les **informations** sur les colonnes (types, valeurs nulles)
4. Affiche les **statistiques descriptives** des colonnes numériques


In [2]:
display(df.head(5))
print("Shape of the dataframe:", df.shape)
print("Number of rows:", df.shape[0], "\nNumber of columns:", df.shape[1])
print("Informations about the data types and nul value:", df.info())
print("Statistics:\n", df.describe())

,vendeur,region,produit,categorie,quantite,prix_unitaire,mois
0,Alice,Nord,Laptop,Tech,3,1200,Janvier
1,Bob,Sud,Chaise,Mobilier,10,150,Janvier
2,Alice,Nord,Ecran,Tech,5,400,Février
3,Clara,Est,Bureau,Mobilier,2,600,Janvier
4,Bob,Sud,Clavier,Tech,8,80,Février


Shape of the dataframe: (12, 7)
Number of rows: 12 
Number of columns: 7
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   vendeur        12 non-null     object
 1   region         12 non-null     object
 2   produit        12 non-null     object
 3   categorie      12 non-null     object
 4   quantite       12 non-null     int64 
 5   prix_unitaire  12 non-null     int64 
 6   mois           12 non-null     object
dtypes: int64(2), object(5)
memory usage: 804.0+ bytes
Informations about the data types and nul value: None
Statistics:
         quantite  prix_unitaire
count  12.000000      12.000000
mean    6.583333     412.500000
std     5.822501     419.954002
min     1.000000      45.000000
25%     2.750000      80.000000
50%     4.500000     275.000000
75%     8.500000     600.000000
max    20.000000    1200.000000


### 1.2 — Exploration rapide

1. Affiche la liste des **noms de colonnes**
2. Affiche le **nombre de ventes par vendeur** (utilise `.value_counts()`)
3. Affiche les **valeurs uniques** de la colonne `"categorie"`


In [3]:
print("Name of the different columns:", df.columns.tolist())
print("Number of sell by vendor:", df["vendeur"].value_counts())
print("Unique values in the categorie column:", df['categorie'].unique())

Name of the different columns: ['vendeur', 'region', 'produit', 'categorie', 'quantite', 'prix_unitaire', 'mois']
Number of sell by vendor: vendeur
Alice    4
Bob      4
Clara    4
Name: count, dtype: int64
Unique values in the categorie column: ['Tech' 'Mobilier']


---

## Partie 2 — Sélection & Filtrage

### 2.1 — Sélectionner des colonnes

1. Affiche uniquement les colonnes `"vendeur"`, `"produit"`, `"prix_unitaire"`
2. Affiche uniquement la colonne `"quantite"` (en tant que Series)
3. Affiche les **3 dernières lignes** du DataFrame (slicing avec `.tail()`)


In [4]:
print(df[["vendeur", "produit", "prix_unitaire"]])
quantity = df['quantite']
print(quantity)
print("Last rows of the dataframe:", df.tail(3))

   vendeur  produit  prix_unitaire
0    Alice   Laptop           1200
1      Bob   Chaise            150
2    Alice    Ecran            400
3    Clara   Bureau            600
4      Bob  Clavier             80
5    Clara   Laptop           1200
6    Alice    Lampe             45
7      Bob    Ecran            400
8    Clara   Chaise            150
9    Alice  Clavier             80
10     Bob   Bureau            600
11   Clara    Lampe             45
0      3
1     10
2      5
3      2
4      8
5      2
6     15
7      3
8      6
9      4
10     1
11    20
Name: quantite, dtype: int64
Last rows of the dataframe:    vendeur region  produit categorie  quantite  prix_unitaire     mois
9    Alice   Nord  Clavier      Tech         4             80     Mars
10     Bob    Sud   Bureau  Mobilier         1            600  Janvier
11   Clara    Est    Lampe  Mobilier        20             45  Février


### 2.2 — Filtrer des lignes

1. Affiche uniquement les ventes de la **région Nord**
2. Affiche uniquement les ventes de catégorie **Tech** avec un `prix_unitaire` **>= 400**
3. Affiche les ventes du vendeur **Alice** ou **Clara** (utilise `|` ou `.isin()`)
4. Affiche les ventes où la `quantite` est **strictement supérieure à la moyenne** des quantités

> 💡 Pour la question 4 : calcule d'abord la moyenne avec `.mean()`, puis utilise-la dans le filtre


In [5]:
display(df.head(5))

,vendeur,region,produit,categorie,quantite,prix_unitaire,mois
0,Alice,Nord,Laptop,Tech,3,1200,Janvier
1,Bob,Sud,Chaise,Mobilier,10,150,Janvier
2,Alice,Nord,Ecran,Tech,5,400,Février
3,Clara,Est,Bureau,Mobilier,2,600,Janvier
4,Bob,Sud,Clavier,Tech,8,80,Février


In [6]:
# Ton code ici
mask = df.query("region == 'Nord'")
print("Vente de la region Nord:\n", mask)
mask_2 = df.query("categorie == 'Tech' and prix_unitaire >= 400")
print(mask_2)
mask_3 = df.query("vendeur == 'Alice' | vendeur == 'Clara'")
print("Vente de Alice ou Clare:\n", mask_3)

Vente de la region Nord:
   vendeur region  produit categorie  quantite  prix_unitaire     mois
0   Alice   Nord   Laptop      Tech         3           1200  Janvier
2   Alice   Nord    Ecran      Tech         5            400  Février
6   Alice   Nord    Lampe  Mobilier        15             45     Mars
9   Alice   Nord  Clavier      Tech         4             80     Mars
  vendeur region produit categorie  quantite  prix_unitaire     mois
0   Alice   Nord  Laptop      Tech         3           1200  Janvier
2   Alice   Nord   Ecran      Tech         5            400  Février
5   Clara    Est  Laptop      Tech         2           1200  Février
7     Bob    Sud   Ecran      Tech         3            400     Mars
Vente de Alice ou Clare:
    vendeur region  produit categorie  quantite  prix_unitaire     mois
0    Alice   Nord   Laptop      Tech         3           1200  Janvier
2    Alice   Nord    Ecran      Tech         5            400  Février
3    Clara    Est   Bureau  Mobilier    

In [7]:
mean_quantity = df['quantite'].mean()
print("Mean quantity:", round(mean_quantity, 2))
mask_4 = df.query("quantite > @mean_quantity")
print(mask_4)

Mean quantity: 6.58
   vendeur region  produit categorie  quantite  prix_unitaire     mois
1      Bob    Sud   Chaise  Mobilier        10            150  Janvier
4      Bob    Sud  Clavier      Tech         8             80  Février
6    Alice   Nord    Lampe  Mobilier        15             45     Mars
11   Clara    Est    Lampe  Mobilier        20             45  Février


---

## Partie 3 — Colonnes calculées & Tri

### 3.1 — Nouvelle colonne

1. Crée une colonne `"chiffre_affaires"` = `quantite * prix_unitaire`
2. Crée une colonne `"ca_categorie"` qui vaut `"CA élevé"` si `chiffre_affaires >= 1000`, sinon `"CA faible"`

> 💡 Pour la question 2 : utilise `np.where(condition, valeur_si_vrai, valeur_si_faux)` — pense à `import numpy as np`


In [8]:
import numpy as np

In [9]:
# Ton code ici
df["chiffre_affaires"] = df["quantite"] * df["prix_unitaire"]
display(df.head(5))
df["ca_categorie"] = np.where(df['chiffre_affaires'] >= 1000, "CA élevé", "CA faible")
display(df.head(15))

,vendeur,region,produit,categorie,quantite,prix_unitaire,mois,chiffre_affaires
0,Alice,Nord,Laptop,Tech,3,1200,Janvier,3600
1,Bob,Sud,Chaise,Mobilier,10,150,Janvier,1500
2,Alice,Nord,Ecran,Tech,5,400,Février,2000
3,Clara,Est,Bureau,Mobilier,2,600,Janvier,1200
4,Bob,Sud,Clavier,Tech,8,80,Février,640


,vendeur,region,produit,categorie,quantite,prix_unitaire,mois,chiffre_affaires,ca_categorie
0,Alice,Nord,Laptop,Tech,3,1200,Janvier,3600,CA élevé
1,Bob,Sud,Chaise,Mobilier,10,150,Janvier,1500,CA élevé
2,Alice,Nord,Ecran,Tech,5,400,Février,2000,CA élevé
3,Clara,Est,Bureau,Mobilier,2,600,Janvier,1200,CA élevé
4,Bob,Sud,Clavier,Tech,8,80,Février,640,CA faible
5,Clara,Est,Laptop,Tech,2,1200,Février,2400,CA élevé
6,Alice,Nord,Lampe,Mobilier,15,45,Mars,675,CA faible
7,Bob,Sud,Ecran,Tech,3,400,Mars,1200,CA élevé
8,Clara,Est,Chaise,Mobilier,6,150,Mars,900,CA faible
9,Alice,Nord,Clavier,Tech,4,80,Mars,320,CA faible


In [10]:
df['ca_categorie'].value_counts()

ca_categorie
CA élevé     6
CA faible    6
Name: count, dtype: int64

### 3.2 — Trier

1. Affiche le DataFrame trié par `chiffre_affaires` **décroissant**
2. Affiche le DataFrame trié par `vendeur` (alphabétique) puis par `chiffre_affaires` décroissant — tri sur **deux colonnes**

> 💡 `.sort_values(["col1", "col2"], ascending=[True, False])`


In [11]:
df = df.sort_values('chiffre_affaires', ascending=False)
display(df.head(12))

,vendeur,region,produit,categorie,quantite,prix_unitaire,mois,chiffre_affaires,ca_categorie
0,Alice,Nord,Laptop,Tech,3,1200,Janvier,3600,CA élevé
5,Clara,Est,Laptop,Tech,2,1200,Février,2400,CA élevé
2,Alice,Nord,Ecran,Tech,5,400,Février,2000,CA élevé
1,Bob,Sud,Chaise,Mobilier,10,150,Janvier,1500,CA élevé
3,Clara,Est,Bureau,Mobilier,2,600,Janvier,1200,CA élevé
7,Bob,Sud,Ecran,Tech,3,400,Mars,1200,CA élevé
11,Clara,Est,Lampe,Mobilier,20,45,Février,900,CA faible
8,Clara,Est,Chaise,Mobilier,6,150,Mars,900,CA faible
6,Alice,Nord,Lampe,Mobilier,15,45,Mars,675,CA faible
4,Bob,Sud,Clavier,Tech,8,80,Février,640,CA faible


In [68]:
df = df.sort_values(['vendeur', 'chiffre_affaires'], ascending=[True, False])
display(df)

,vendeur,region,produit,categorie,quantite,prix_unitaire,mois,chiffre_affaires,ca_categorie
0,Alice,Nord,Laptop,Tech,3,1200,Janvier,3600,CA élevé
2,Alice,Nord,Ecran,Tech,5,400,Février,2000,CA élevé
6,Alice,Nord,Lampe,Mobilier,15,45,Mars,675,CA faible
9,Alice,Nord,Clavier,Tech,4,80,Mars,320,CA faible
1,Bob,Sud,Chaise,Mobilier,10,150,Janvier,1500,CA élevé
7,Bob,Sud,Ecran,Tech,3,400,Mars,1200,CA élevé
4,Bob,Sud,Clavier,Tech,8,80,Février,640,CA faible
10,Bob,Sud,Bureau,Mobilier,1,600,Janvier,600,CA faible
5,Clara,Est,Laptop,Tech,2,1200,Février,2400,CA élevé
3,Clara,Est,Bureau,Mobilier,2,600,Janvier,1200,CA élevé


---

## Partie 4 — GroupBy & Agrégations

### 4.1 — GroupBy simple

1. Calcule le **chiffre d'affaires total par vendeur** (`.groupby().sum()`)
2. Calcule la **quantité moyenne vendue par catégorie** (`.groupby().mean()`)
3. Calcule le **nombre de ventes par région** (`.groupby().count()`)


In [70]:
# Ton code ici
ca_vendeur = df.groupby('vendeur')['chiffre_affaires'].sum()
display(ca_vendeur)

vendeur
Alice    6595
Bob      3940
Clara    5400
Name: chiffre_affaires, dtype: int64

In [71]:
mean_categorie_df = df.groupby('categorie')['quantite'].mean()
display(mean_categorie_df)

categorie
Mobilier    9.000000
Tech        4.166667
Name: quantite, dtype: float64

In [15]:
vente_region = df.groupby('region').count()
display(vente_region)

,vendeur,produit,categorie,quantite,prix_unitaire,mois,chiffre_affaires,ca_categorie
region,,,,,,,,
Est,4,4,4,4,4,4,4,4
Nord,4,4,4,4,4,4,4,4
Sud,4,4,4,4,4,4,4,4


### 4.2 — .agg() — plusieurs métriques en une fois

En une seule opération `.groupby().agg()`, calcule **par vendeur** :
- La **somme** du `chiffre_affaires`
- La **moyenne** du `prix_unitaire`
- Le **max** de la `quantite`

Stocke le résultat dans `stats_vendeurs` et affiche-le.

> 💡 `.agg({"col1": "sum", "col2": "mean", "col3": "max"})`


In [16]:
df_sorted = df.groupby('vendeur').agg(
    sum_ca=('chiffre_affaires', 'sum'),
    mean_unit_price=('prix_unitaire', 'mean'),
    max_qty=('quantite', 'max')
)
print(df_sorted)

         sum_ca  mean_unit_price  max_qty
vendeur                                  
Alice      6595           431.25       15
Bob        3940           307.50       10
Clara      5400           498.75       20


### 4.3 — GroupBy multi-niveaux

1. Calcule le **CA total par vendeur ET par mois** (groupby sur deux colonnes)
2. Calcule par **catégorie** : la somme du CA, la quantité totale, et le prix unitaire moyen — avec `.agg()`

> 💡 Pour grouper sur deux colonnes : `df.groupby(["col1", "col2"])`


In [17]:
ca_total_v_m = df.groupby(['vendeur', 'mois']).agg(
    CA_total=('chiffre_affaires', 'sum')
)
print(ca_total_v_m)

                 CA_total
vendeur mois             
Alice   Février      2000
        Janvier      3600
        Mars          995
Bob     Février       640
        Janvier      2100
        Mars         1200
Clara   Février      3300
        Janvier      1200
        Mars          900


In [18]:
sort_categorie = df.groupby('categorie').agg(
    CA=('chiffre_affaires', 'sum'),
    qty_total=('quantite', 'sum'),
    prix_unitaire_moyen=('prix_unitaire', 'mean')
)
print(sort_categorie)

              CA  qty_total  prix_unitaire_moyen
categorie                                       
Mobilier    5775         54                265.0
Tech       10160         25                560.0


---

## 🎯 Bonus

1. Quel vendeur a généré le **chiffre d'affaires total le plus élevé** ? Affiche son nom et son CA. (utilise `groupby` + `idxmax()`)
2. Quel est le **produit le plus vendu** en quantité totale ?
3. Affiche le **top 3 des ventes** (lignes) par chiffre d'affaires décroissant — uniquement les colonnes `"vendeur"`, `"produit"`, `"chiffre_affaires"`

> 💡 Pour question 1 : `.idxmax()` retourne l'index de la valeur maximale dans une Series


In [72]:
ca_par_vendeur = df.groupby('vendeur')['chiffre_affaires'].sum()
meilleur = ca_par_vendeur.idxmax()
print(f"{meilleur} : {ca_par_vendeur[meilleur]} €")

Alice : 6595 €


1. Alice à le chiffre d'affaire le plus elvé, 6595 contre 3940 et 5400 pour Bob et Clara respectivement.

In [64]:
produit_plus_vendu = df.groupby('produit').agg(qty_total=('quantite', 'sum')).sort_values(by='qty_total', ascending=False)
display(produit_plus_vendu)

,qty_total
produit,
Lampe,35
Chaise,16
Clavier,12
Ecran,8
Laptop,5
Bureau,3


In [ ]:
display(df[['vendeur', 'produit', 'chiffre_affaires']].head(3))

,vendeur,produit,chiffre_affaires
0,Alice,Laptop,3600
2,Alice,Ecran,2000
6,Alice,Lampe,675
